programs/whirlpool/src/instructions/v2/swap.rs

```diff
@@ -0,0 +1,262 @@
+use anchor_lang::prelude::*;
+use anchor_spl::token_interface::{Mint, TokenAccount, TokenInterface};
+use anchor_spl::memo::Memo;
+
+use crate::util::{calculate_transfer_fee_excluded_amount, calculate_transfer_fee_included_amount, parse_remaining_accounts, AccountsType, RemainingAccountsInfo};
+use crate::{
+    errors::ErrorCode,
+    manager::swap_manager::*,
+    state::{TickArray, Whirlpool},
+    util::{to_timestamp_u64, v2::update_and_swap_whirlpool_v2, SwapTickSequence},
+    constants::transfer_memo,
+};
+
+#[derive(Accounts)]
+pub struct SwapV2<'info> {
+    #[account(address = token_mint_a.to_account_info().owner.clone())]
+    pub token_program_a: Interface<'info, TokenInterface>,
+    #[account(address = token_mint_b.to_account_info().owner.clone())]
+    pub token_program_b: Interface<'info, TokenInterface>,
+
+    pub memo_program: Program<'info, Memo>,
+
+    pub token_authority: Signer<'info>,
+
+    #[account(mut)]
+    pub whirlpool: Box<Account<'info, Whirlpool>>,
+
+    #[account(address = whirlpool.token_mint_a)]
+    pub token_mint_a: InterfaceAccount<'info, Mint>,
+    #[account(address = whirlpool.token_mint_b)]
+    pub token_mint_b: InterfaceAccount<'info, Mint>,
+    
+    #[account(mut, constraint = token_owner_account_a.mint == whirlpool.token_mint_a)]
+    pub token_owner_account_a: Box<InterfaceAccount<'info, TokenAccount>>,
+    #[account(mut, address = whirlpool.token_vault_a)]
+    pub token_vault_a: Box<InterfaceAccount<'info, TokenAccount>>,
+
+    #[account(mut, constraint = token_owner_account_b.mint == whirlpool.token_mint_b)]
+    pub token_owner_account_b: Box<InterfaceAccount<'info, TokenAccount>>,
+    #[account(mut, address = whirlpool.token_vault_b)]
+    pub token_vault_b: Box<InterfaceAccount<'info, TokenAccount>>,
+
+    #[account(mut, has_one = whirlpool)]
+    pub tick_array_0: AccountLoader<'info, TickArray>,
+
+    #[account(mut, has_one = whirlpool)]
+    pub tick_array_1: AccountLoader<'info, TickArray>,
+
+    #[account(mut, has_one = whirlpool)]
+    pub tick_array_2: AccountLoader<'info, TickArray>,
+
```

Comment:
Yep, I believe it is okay for now.

In my current thought, we need to change it to UncheckedAccount (unfortunately) in the next cycle, because we need to accept uninitialized account address to check account is really uninitialized.
In my understanding for Option implementation in Anchor, it uses ProgramId as padding.
So Option account can accept only ProgramID or initialized account.

https://github.com/coral-xyz/anchor/pull/2101

>How this works?
>
>The key to make everything work is the fact that repeated accounts don't add to the size of the transaction due to how Solana serializes accounts in transactions. We use this fact to denote an optional account with the program Id.
>
> When the accounts are passed in, if an account should be ignored, clients will send the program Id instead.
>
> In addition, with this implementation, adding on optional accounts at the end of an Accounts struct wouldn't be a breaking change and the program will treat the missing account as None (since older clients will not be passing in this new account).